File: Model.UnitTest/unittest1.cpp
#include "stdafx.h"
#include "CppUnitTest.h"

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace ModelUnitTest
{
	TEST_CLASS(UnitTest1)
	{
	public:

		TEST_METHOD(ModelCreate)
		{
			Model::TreeNode* node = CreateTestModel();
			Assert::IsNotNull(node);
			delete node;
		}

		TEST_METHOD(ModelRemoveFirstChild)
		{
			Model::TreeNode* node = CreateTestModel();
			Model::TreeNode* firstChild = node->GetFirstChild();
			node->RemoveChild(firstChild);
			Assert::IsNull(firstChild->GetParent());
			delete firstChild;
			delete node;
		}

		TEST_METHOD(ModelSerialization)
		{
			const wchar_t* fileName = L"model.bin";
			Model::TreeNode* node = CreateTestModel();
			node->Serialize(fileName);
			Model::TreeNode* node1 = Model::TreeNode::Deserialize(fileName);
			Assert::IsTrue(node->GetValue() == node1->GetValue());
			Assert::IsTrue(node->GetFirstChild()->GetValue() == node1->GetFirstChild()->GetValue());
			Assert::IsTrue(node->GetLastChild()->GetValue() == node1->GetLastChild()->GetValue());
			delete node;
			delete node1;
		}

	private:
		Model::TreeNode* CreateTestModel()
		{
			Model::TreeNode* root = new Model::TreeNode(4);

			Model::TreeNode* node1 = new Model::TreeNode(8);
			root->AddChild(node1);

			Model::TreeNode* node11 = new Model::TreeNode(7);
			node1->AddChild(node11);

			Model::TreeNode* node12 = new Model::TreeNode(1);
			node1->AddChild(node12);

			Model::TreeNode* node2 = new Model::TreeNode(2);
			root->AddChild(node2);

			Model::TreeNode* node21 = new Model::TreeNode(9);
			node2->AddChild(node21);

			return root;
		}

	};
}

File: ModelWrapper.UnitTest/UnitTest1.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using ViewModel;

namespace ModelWrapper.UnitTest
{
	[TestClass]
	public class UnitTest1
	{
		[TestMethod]
		public void WrapperCreate()
		{
			TreeNode node = CreateTestModel();
			Assert.IsNotNull( node );
		}

		[TestMethod]
		public void WrapperRemoveFirstChild()
		{
			TreeNode node = CreateTestModel();
			TreeNode firstChild = node.Children[0];
			node.RemoveChild( firstChild );
			Assert.IsNull( firstChild.Parent );
		}

		[TestMethod]
		public void WrapperSerialization()
		{
			string fileName = "wrapper.bin";
			TreeNode node = CreateTestModel();
			node.Save( fileName );
			TreeNode node1 = TreeNode.Open( fileName );
			Assert.IsTrue( node.Value == node1.Value );
			Assert.IsTrue( node.Children[0].Value == node1.Children[0].Value );
			Assert.IsTrue( node.Children[node.Children.Count - 1].Value == node1.Children[node1.Children.Count - 1].Value );
		}

		[TestMethod]
		public void ViewModelCreate()
		{
			TreeNode node = CreateTestModel();
			TreeNodeVM vm = new TreeNodeVM( node );
			Assert.IsNotNull( vm );
		}

		[TestMethod]
		public void ViewModelRemoveFirstChild()
		{
			TreeNode node = CreateTestModel();
			TreeNodeVM vm = new TreeNodeVM( node );
			TreeNodeVM firstChild = vm.Children[0];
			vm.Children.RemoveAt( 0 );
			Assert.IsNull( firstChild.Parent );
		}

		[TestMethod]
		public void ViewModelUpdateValue()
		{
			TreeNode node = CreateTestModel();
			TreeNodeVM vm = new TreeNodeVM( node );
			TreeNodeVM firstChild = vm.Children[0];
			firstChild.Value = 17;
			Assert.IsTrue( firstChild.Value == firstChild.Node.Value );
		}


		private TreeNode CreateTestModel()
		{
			TreeNode root = new TreeNode( 4 );

			TreeNode node1 = new TreeNode( 8 );
			root.AddChild( node1 );

			TreeNode node11 = new TreeNode( 7 );
			node1.AddChild( node11 );

			TreeNode node12 = new TreeNode( 1 );
			node1.AddChild( node12 );

			TreeNode node2 = new TreeNode( 2 );
			root.AddChild( node2 );

			TreeNode node21 = new TreeNode( 9 );
			node2.AddChild( node21 );

			return root;
		}
	}
}
